modules/040-node-manager/hooks/fencing_controller.go

```diff
@@ -0,0 +1,166 @@
+		{
+			Name:       nodesSnapshot,
+			ApiVersion: "v1",
+			Kind:       "Node",
+			LabelSelector: &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      "node-manager.deckhouse.io/fencing-enabled",
+						Operator: metav1.LabelSelectorOpExists,
+					},
+				},
+			},
+			FilterFunc: fencingControllerNodeFilter,
+		},
+	},
+}, dependency.WithExternalDependencies(fencingControllerHandler))
+
+type fencingControllerNodeResult struct {
+	Name          string
+	NodeGroupName string
+}
+
+type fencingControllerLeaseResult struct {
+	NodeName  string
+	RenewTime time.Time
+}
+
+func fencingControllerNodeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var res fencingControllerNodeResult
+
+	for _, annotation := range maintenanceAnnotations {
+		_, annotationExists := obj.GetAnnotations()[annotation]
+		if annotationExists {
+			return nil, nil
+		}
+	}
+
+	res.Name = obj.GetName()
+	return res, nil
+}
+
+func fencingControllerHandler(input *go_hook.HookInput, dc dependency.Container) error {
+	if len(input.Snapshots[nodesSnapshot]) == 0 {
+		// No nodes with enabled fencing -> nothing to do
+		return nil
+	}
+
+	// kubeclient to get node leases and get and delete pods
+	kubeClient, err := dc.GetK8sClient()
+	if err != nil {
+		input.LogEntry.Errorf("%v", err)
+		return err
+	}
+
+	// make map with nodes to kill
+	nodesToKill := set.New()
+	for _, nodeRaw := range input.Snapshots[nodesSnapshot] {
+		if nodeRaw == nil {
+			continue
+		}
+
+		node := nodeRaw.(fencingControllerNodeResult)
+		nodeLease, err := kubeClient.CoordinationV1().Leases("kube-node-lease").Get(context.TODO(), node.Name, metav1.GetOptions{})
+		if err != nil {
+			input.LogEntry.Errorf("Can't get node lease: %v", err)
+			continue
+		}
+
+		if time.Since(nodeLease.Spec.RenewTime.Time) > fencingControllerTimeout {
+			input.LogEntry.Warnf("Node lease %s is expired. Current time: %v, node lease time %v", node.Name, time.Now(), nodeLease.Spec.RenewTime.Time)
+			nodesToKill.Add(node.Name)
+		}
+	}
+
+	nodeToKillCount := nodesToKill.Size()
+	if nodeToKillCount == 0 {
+		// nothing to kill -> skip
+		return nil
+	}
+
+	input.LogEntry.Warnf("Going to kill %d nodes", nodeToKillCount)
+
+	// kill nodes
+	for _, node := range nodesToKill.Slice() {
+		input.LogEntry.Warnf("Delete all pods from node %s", node)
+		podsToDelete, err := kubeClient.CoreV1().Pods("").List(
+			context.TODO(),
+			metav1.ListOptions{
+				FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": node}).String(),
+			},
+		)
+		if err != nil {
+			input.LogEntry.Errorf("Can't list pods: %v", err)
+			continue
+		}
+
+		GracePeriodSeconds := int64(0)
+		for _, pod := range podsToDelete.Items {
+			input.LogEntry.Warnf("Delete pod %s in namespace %s on node %s", pod.Name, pod.Namespace, pod.Spec.NodeName)
+			err = kubeClient.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
```

Comment:
Do we need cordon node before deleting pods?

Reply:
I could to do deleting pods in parallel

Reply:
Or maybe use drain controller https://github.com/deckhouse/deckhouse/blob/0f1520c4b38a1e1bfac5e2088d0149bb68aa3890/modules/040-node-manager/hooks/handle_draining.go#L172

Reply:
Will this work correctly if the node is not available over the network?

Reply:
Under hood it deletes pods https://github.com/deckhouse/deckhouse/blob/main/go_lib/dependency/k8s/drain/drain.go#L143 similary, but it is wait for pod will be evicted.

In our case with unavailable node, maybe, better way to delete pods simplify